### PR TITLE
Allows viewing event log and sending emails to end users when samples are scanned

### PR DESCRIPTION
**Pull request by dhakim87**
Not the prettiest, but seems functional.  There is no warning or confirmation when you send emails, so be careful to fill out the form completely for now.

### CODE REVIEW

**Comment by wasade** [comment]
This seems good, haz screenshot by chance? Travis looks like it got angry about a change on a route.

**Comment by wasade** [comment]
@AmandaBirmingham, since you're working on the scanning page in parallel right now, would it be possible to do a review here? 

On slack, @dhakim87 commented that it may be good to require a technician click a confirmation button before sending the email, or to at least present the content of the email to the technician. I'm not sure what's best here, curious about thoughts

**Comment by dhakim87** [comment]
Review with Dom and Kat:

- Based on sample status, autofill as much as possible.  
- Send email based on clicking the add scan button, so long as an email needs to be sent - full policy is tbd, but should be similar to “status has changed and checkbox is checked and all information necessary to send email template is available”
- Keep a checkbox determining whether to send email on the page (What are rules for persistence of this between samples/sessions?)
- When sample status changes, (and checkbox is unchecked), send email dependent on new status.

**Comment by AmandaBirmingham** [comment]
@dhakim87 Interesting feedback from the Dom/Kat review!  Can you please talk with MacKenzie in the wet lab, if you haven't already? AFAIK she is most likely to be the person actually clicking the button that sends an email, and in earlier discussions of the barcode-scanning-related functionality, she had some requirements that surprised me (like not wanting the sample status dropdown set to what we knew the status of the sample was in the db when there was a problem with it, but instead wanting the wet lab personnel to be required to change it manually ...)

**Comment by dhakim87** [comment]
Checked the email chain, MacKenzie's feedback was included, Kat was not in the call.  My mistake.

**Comment by dhakim87** [comment]
In the next review, please give special attention to ruleset.js, it's my attempt at a framework to simplify implementing business logic on client side by letting you separate out, cache, name, and combine the jquery $().change(...) calls.  It could definitely use some more layers of syntactic sugar, but it helped a heck of a lot for simplifying enabling/disabling, showing/hiding, changing text, and everything else.  

I used it heavily in implementing the feedback from Dom and MacKenzie.

File: microsetta_admin/templates/scan.html [hunk 1]
```diff
@@ -1,4 +1,140 @@
+{% set SAMPLE_DROPDOWN = [dummy_status, "Blood (skin prick)", "Saliva", "Ear wax", "Forehead", "Fur", "Hair", "Left hand", "Left leg", "Mouth", "Nares", "Nasal mucus",
```
Comment (wasade) [inline microsetta_admin/templates/scan.html:1]: cringe...

File: microsetta_admin/templates/scan.html [hunk 2]
```diff
@@ -122,6 +268,31 @@ <h3>Microsetta Scan</h3>
                 </select>
                 </td>
             </tr>
+            <tr>
+                <td>Send Email Upon Status Update:</td>
+                <td><input type="checkbox" id="send_email" name="send_email" {% if send_email %}checked{% endif %}/><label for="send_email"><small> <b><i><span id="send_email_label_text"></span></i></b></small></label></td>
+            </tr>
+            <!-- Begin collapsible email template fields -->
```
Comment (wasade) [inline microsetta_admin/templates/scan.html:275]: collapsable

File: microsetta_admin/tests/test_routes.py [hunk 3]
```diff
@@ -42,37 +46,63 @@ def test_scan_simple(self):
...
+        dynObj2.text = json.dumps(resp2)
+        dynObj2.json = lambda: resp2
+
+        self.mock_get.side_effect = [dynObj, dynObj2]
 
         response = self.app.get('/scan?sample_barcode=000004216',
                                 follow_redirects=True)
 
         self.assertEqual(response.status_code, 200)
+        print(response.status_code)
```
Comment (wasade) [inline microsetta_admin/tests/test_routes.py:77]: remove please

File: microsetta_admin/server.py [hunk 4]
```diff
@@ -326,14 +341,80 @@ def scan():
...
+                "template": template,
+                "template_args": {
+                    "sample_barcode": sample_barcode,
+                    "recorded_type": recorded_type,
+                    "received_type": received_type
+                }
+            }
+        )
+
+        # if the update failed, keep track of the error so it can be displayed
```
Comment (AmandaBirmingham) [inline microsetta_admin/server.py:376]: Could comment say "if emailing failed", or something?

File: microsetta_admin/templates/scan.html (hunk 2 above)
Comment (dhakim87) [reply microsetta_admin/templates/scan.html:290]: Collapsible is valid too :P

File: microsetta_admin/server.py [hunk 5]
```diff
@@ -255,12 +256,12 @@ def _check_sample_status(extended_barcode_info):
...
 
-    # If its a get, grab the sample_barcode from the query string rather than
-    # form parameters
-    if request.method == 'GET':
-        sample_barcode = request.args.get('sample_barcode')
+    # Set up handlers for the cases,
+    #   GET to view the page,
+    #   POST to update info for a barcode -AND (possibly)-
+    #        email end user about the change in sample status,
+    def _scan_get(sample_barcode, update_error):
```
[the remark this answers is not in the document]
Comment (dhakim87) [reply microsetta_admin/server.py:264]: Slightly reduces scoping I guess?  Any reason not to?  There had been two separate _scan_post helper methods with different behaviors prior to this PR, and the _scan_get is used by the _scan_post, so keeping a separate GET and POST method at the top level was leaving redundant code.

File: microsetta_admin/templates/scan.html (hunk 2 above)
Comment (wasade) [reply microsetta_admin/templates/scan.html:290]: did not know that... ah! sorry :)

**Comment by dhakim87** [comment]
For a slightly simpler diff, merge the unit test PR prior to this one.

**Comment by wasade** [comment]
:+1: pending comment from our call

**Pull request merged**